crates/typos-cli/src/bin/typos-cli/report.rs

```diff
@@ -280,6 +286,155 @@ impl Report for PrintJson {
+                        )
+                        .build().map_err(error_to_io_error)?)
+                    .build().map_err(error_to_io_error)?;
+
+
+                let result = sarif::ResultBuilder::default()
+                    .level(sarif::ResultLevel::Error.to_string())
+                    .message(sarif::MessageBuilder::default()
+                        .markdown(message)
+                        .build().map_err(error_to_io_error)?)
+                    .locations(vec![location])
+                    .build().map_err(error_to_io_error)?;
+
+                self.results.lock().unwrap().borrow_mut().push(result);
+            }
+            Message::Parse(msg) => {
+                self.error.lock().unwrap().borrow_mut().push(msg.data.to_owned());
+            }
+            Message::Error(msg) => {
+                self.error.lock().unwrap().borrow_mut().push(msg.msg.clone());
+            }
+            Message::BinaryFile(_) => {}
+            Message::FileType(_) => {}
+            Message::File(_) => {}
+            _ => unimplemented!("New message {:?}", msg),
+        }
+
+        Ok(())
+    }
+
+    fn finalize(&self) -> Result<(), std::io::Error> {
+        let mut sarif_builder = sarif::SarifBuilder::default();
+        sarif_builder.version(sarif::Version::V2_1_0.to_string())
+            .schema(sarif::SCHEMA_URL);
+
+
+        let tool = sarif::ToolBuilder::default()
+            .driver(sarif::ToolComponentBuilder::default()
+                .name("typos")
+                .information_uri("https://github.com/crate-ci/typos")
+                .build().map_err(error_to_io_error)?)
+            .build().map_err(error_to_io_error)?;
+
+        let mut run_builder = sarif::RunBuilder::default();
+        run_builder
+            .tool(tool)
+            .column_kind(sarif::ResultColumnKind::UnicodeCodePoints.to_string())
+            .results(self.results.lock().unwrap().borrow().clone());
+
+
+        if self.error.lock().unwrap().borrow().len() > 0 {
+            let invocations: Vec<Result<Invocation, std::io::Error>> = self.error.lock().unwrap().borrow().iter().map(
+                |x| {
+                    sarif::InvocationBuilder::default()
+                        .process_start_failure_message(x.to_owned().clone())
+                        .build().map_err(error_to_io_error)
+                }
+            ).collect();
+            
+            let error = invocations.iter().find(|x| x.is_err());
+            // if invocations contains any error, return error
+            if error.is_some() {
+                let error = error.unwrap().as_ref().unwrap_err();
+                return Err(std::io::Error::new(error.kind(), error.to_string()));
+            }
+            
+            let invocations: Vec<Invocation> = invocations.into_iter().map(|x| x.unwrap()).collect();
+
+            run_builder.invocations(
+                invocations
+            );
+        }
```

Comment:
It looks like this 
1. Converts the error
2. Checks for the error
3. Unwraps because the error is handled

Generally avoid checking for an error and then unwrapping.  Most likely, the culprit for this is over-reliance on combinators.

Reply:
How about https://github.com/crate-ci/typos/pull/1047/commits/5ad1fb9e130db2d5ebc9c0bc5699413bbf45ea7b ?

Reply:
Thanks, that cleans it up